circuit/block_constraints.go

```diff
@@ -99,9 +111,37 @@ func VerifyBlock(
 			count++
 		}
 		onChainOpsCount = api.Add(onChainOpsCount, isOnChainOp)
+		for i := 0; i < gasAssetCount; i++ {
+			for j := 0; j < NbGasAssetsPerTx; j++ {
+				found := api.IsZero(api.Sub(block.GasAssetIds[i], gasDeltas[j].AssetId))
+				delta := api.Select(found, gasDeltas[j].AssetId, types.ZeroInt)
+				blockGasDeltas[i] = api.Add(blockGasDeltas[i], delta)
+			}
+		}
 	}
+
+	needGas = 0
+	for i := 0; i < block.TxsCount; i++ {
+		transferTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeTransfer))
+		withdrawTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeWithdraw))
+		createCollectionTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeCreateCollection))
+		mintNftTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeMintNft))
+		cancelOfferTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeCancelOffer))
+		atomicMatchTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeAtomicMatch))
+		withdrawNftTx := api.IsZero(api.Sub(block.Txs[i].TxType, types.TxTypeWithdrawNft))
+		needGas = api.Or(api.Or(api.Or(api.Or(api.Or(api.Or(transferTx, withdrawTx), createCollectionTx), mintNftTx), cancelOfferTx), atomicMatchTx), withdrawNftTx)
+	}
+
+	newAccountRoot, err := VerifyGas(api, block.Gas, needGas, blockGasDeltas, hFunc, roots[0])
+	hFunc.Reset()
+	hFunc.Write(
+		newAccountRoot,
+		roots[1],
+	)
+	newStateRoot := hFunc.Sum()
+	types.IsVariableEqual(api, needGas, block.NewStateRoot, newStateRoot)
```

Comment:
When `needGas` is false, we should make sure that block.NewStateRoot equals to last meaningful `stateRoot`